ts2mls/js/src/main/scala/ts2mls/TSCompilerInfo.scala

```diff
@@ -0,0 +1,170 @@
+class TSTypeChecker(checker: js.Dynamic) {
+  def getTypeStringOfSymbol(sym: js.Dynamic): String =
+    checker.typeToString(getTypeOfSymbolAtLocation(sym, sym.valueDeclaration)).toString
+
+  def getReturnTypeOfSignature(node: js.Dynamic) = checker.getReturnTypeOfSignature(checker.getSignatureFromDeclaration(node))
+  def getTypeFromTypeNode(node: js.Dynamic) = checker.getTypeFromTypeNode(node)
+  def getTypeOfSymbolAtLocation(sym: js.Dynamic, node: js.Dynamic) = checker.getTypeOfSymbolAtLocation(sym, node)
+  def getPropertiesOfType(tp: js.Dynamic) = checker.getPropertiesOfType(tp)
+
+  def isImplementationOfOverload(node: js.Dynamic) = checker.isImplementationOfOverload(node)
+  def typeToTypeNode(tp: js.Dynamic) = checker.typeToTypeNode(tp)
+  def getTypeArguments(tp: js.Dynamic) = checker.getTypeArguments(tp)
+  def getElementTypeOfArrayType(tp: js.Dynamic) = checker.getElementTypeOfArrayType(tp)
+}
+
+object TSTypeChecker {
+  def apply(checker: js.Dynamic) = new TSTypeChecker(checker)
+}
+
+class TSSymbolObject(sym: js.Dynamic)(implicit checker: TSTypeChecker) extends TSAny(sym) {
+  private lazy val parent = TSSymbolObject(sym.parent)
+
+  // the first declaration of this symbol
+  // if there is no overloading, there is only one declaration
+  lazy val declaration =
+    if (declarations.isUndefined) TSNodeObject(g.undefined)
+    else declarations.get(0)
+  lazy val escapedName: String = sym.escapedName.toString
+  lazy val declarations = TSNodeArray(sym.declarations)
+
+  lazy val symbolType: String = checker.getTypeStringOfSymbol(sym)
+
+  // get the full name of the reference symbol
+  // e.g. class A extends B => class A extends SomeNamespace'B
+  lazy val fullName: String =
+    if (parent.isUndefined || !parent.declaration.isNamespace) escapedName
+    else s"${parent.fullName}'$escapedName"
+}
+
+object TSSymbolObject {
+  def apply(node: js.Dynamic)(implicit checker: TSTypeChecker) = new TSSymbolObject(node)
+}
+
+class TSNodeObject(node: js.Dynamic)(implicit checker: TSTypeChecker) extends TSAny(node) {
+  private lazy val modifiers = TSTokenArray(node.modifiers)
+
+  lazy val isToken = TypeScript.isToken(node)
+  lazy val isClassDeclaration = TypeScript.isClassDeclaration(node)
+  lazy val isInterfaceDeclaration = TypeScript.isInterfaceDeclaration(node)
+  lazy val isFunctionLike = TypeScript.isFunctionLike(node)
+
+  // `TypeScript.isModuleDeclaration` works on both namespaces and modules
+  // but namespaces are more recommended, so we merely use `isNamespace` here
+  lazy val isNamespace = TypeScript.isModuleDeclaration(node)
+  lazy val isArrayTypeNode = TypeScript.isArrayTypeNode(node)
+  lazy val isTupleTypeNode = TypeScript.isTupleTypeNode(node)
+  lazy val isImplementationOfOverload = checker.isImplementationOfOverload(node)
+
+  // if a node has an initializer or is marked by a question notation
+  // it is ok if we provide nothing
+  lazy val isOptional = !initializer.isUndefined || !IsUndefined(node.questionToken)
```

Comment:
What does "we provide nothing" mean here?

Also, why the `!initializer.isUndefined` check?

Reply:
ohh! some dumbass comments. :dizzy_face: I've updated it.

Reply:
e.g. `function f(x?: int) {}`, we can use it directly: `f()`.
in this case, there would be a question token.
e.g. `function f(x: int = 42) {}`, we can use it directly: `f()`.
in this case, the initializer would store the default value and be non-empty.

Reply:
I see. This makes me wonder why we don't directly look at the _type_ of the definition, as determined by TS, instead of looking at the definition tree. TS will resolve all these details on its own already. Can't we just query the type of `f` here?

Reply:
Actually, I've tried to use `TypeObject` in the code as much as possible, and I think that is the information of type determined by TSC.

Reply:
So why not use it here too? Do you remember any specific problem while doing so? If possible that would be better, as it would make the code less brittle.

Reply:
ok I will try. 😸

Reply:
I just found another two API that can be used to check optional parameters and optional members in interfaces seperately. These operations are not manipulated on `TypeObject` but I think they can  at least improve the robuness of codes.

Reply:
Cool 👍